src/NHibernate/AdoNet/ConnectionManager.cs

```diff
@@ -270,12 +281,17 @@ public void AfterStatement()
 
 		private void AggressiveRelease()
 		{
-			if (ownConnection && flushingFromDtcTransaction == false)
+			if (_ownConnection)
 			{
-				log.Debug("aggressively releasing database connection");
-				if (connection != null)
+				if (_connection != null)
 				{
-					CloseConnection();
+					if (_processingFromSystemTransaction)
+						_connectionReleaseRequired = true;
```

Comment:
Previously `flushingFromDtcTransaction` was used to prevent releasing the connection during the flush from transaction prepare phase. But since some times the `IsAgressiveRelease` condition is in-fact checking the `IsInActiveTransaction` flag of the session context, causing this `flushingFromDtcTransaction` to be indeed useless. The `AgressiveRelease` is no more called from `AfterStatement` when a system transaction is ongoing, even when called from another thread. (This could change if instead of checking `IsInActiveTransaction` flag the code is changed for checking `System.Transactions.Transaction.Current` instead, which would missed calls from other threads, such as those from MSDTC.)

So now we can use this flag for detecting the `AfterTransaction` trying to close the connection from system transaction, and then just flag the connection for releasing instead of releasing it there.